### PR TITLE
Add a repodata gateway, refactoring existing code

### PR DESCRIPTION
**Pull request by dholth**
### Description

Adapt features/repo-plugin to pass tests, separate jlap support from refactor.

### Checklist - did you ...

- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

File: conda/core/repo.py [hunk 1]
```diff
@@ -1,261 +1,277 @@
...
     RequestsProxyError,
+    SSLError,
 )
 from conda.gateways.connection.session import CondaSession
 from conda.models.channel import Channel
 
+log = logging.getLogger(__name__)
+stderrlog = logging.getLogger("conda.stderrlog")
 
-Repodata = TypedDict('Repodata', {  # TODO: Use more specific types
```
Comment (dholth) [inline conda/core/repo.py]: not available in Python 3.7

File: tests/core/test_subdir_data.py [hunk 2]
```diff
@@ -173,19 +204,18 @@ def test_cache_fn_url_repo_anaconda_com(self):
 
 
 class FetchLocalRepodataTests(TestCase):
-
     def test_fetch_repodata_remote_request_invalid_arch(self):
         # see https://github.com/conda/conda/issues/8150
-        url = 'file:///fake/fake/fake/linux-64'
+        url = "file:///fake/fake/fake/linux-64"
```
Comment (dholth) [inline tests/core/test_subdir_data.py:209]: I wonder if this is real

File: tests/test_create.py [hunk 3]
```diff
@@ -136,20 +136,20 @@ def test_run_preserves_arguments(self):
                     assert args[i] == line.replace('\r', '')
 
     def test_create_install_update_remove_smoketest(self):
-        with make_temp_env("python=3.5") as prefix:
+        with make_temp_env("python=3.9") as prefix:
```
Comment (dholth) [inline tests/test_create.py:139]: avoid versions missing from newer repositories

File: tests/requirements.txt [hunk 4]
```diff
@@ -9,11 +9,13 @@ conda-build
 conda-forge::pre-commit
 conda-forge::pytest-split
 conda-forge::pytest-xprocess
+conda-forge::werkzeug >=2.2
 conda-forge::xdoctest
 conda-package-handling
 conda-verify
 filelock
 flake8
+conda-forge::flask >=2.2
```
Comment (dholth) [inline tests/requirements.txt:18]: will be necessary for jlap tests

File: tests/core/test_subdir_data.py [hunk 5]
```diff
@@ -203,31 +233,34 @@ def CondaSession_get(*args, **kwargs):
         fetch_repodata_remote_request(url, etag, mod_stamp)
 
 
-def test_subdir_data_prefers_conda_to_tar_bz2():
-    channel = Channel(join(dirname(__file__), "..", "data", "conda_format_repo", context.subdir))
+def test_subdir_data_prefers_conda_to_tar_bz2(platform="linux-64"):
```
Comment (dholth) [inline tests/core/test_subdir_data.py:236]: necessary to pass on platforms that don't exist in our test in-conda-source-tree repository

File: conda/core/subdir_data.py [hunk 6]
```diff
@@ -170,16 +154,27 @@ def __init__(self, channel, repodata_fn=REPODATA_FN):
...
+        self.url_w_subdir = self.channel.url(with_credentials=False) or ""
+        self.url_w_credentials = self.channel.url(with_credentials=True) or ""
         # whether or not to try using the new, trimmed-down repodata
         self.repodata_fn = repodata_fn
         self._loaded = False
         self._key_mgr = None
 
+    @property
+    def _repo(self) -> RepoInterface:
+        # XXX one for `current_repodata.json` and one for `repodata.json`
```
Comment (dholth) [inline conda/core/subdir_data.py:170]: The class changes its url when retrying repodata.json, when current_repodata.json fails to load.

File: tests/core/test_subdir_data.py [hunk 7]
```diff
@@ -203,31 +233,34 @@ def CondaSession_get(*args, **kwargs):
...
     # force this to False, because otherwise tests fail when run with old conda-build
```
Comment (dholth) [inline tests/core/test_subdir_data.py:237]: this "run with old conda-build" thing may not be a problem anymore

File: conda/core/subdir_data.py (hunk 6 above)
Comment (dholth) [reply conda/core/subdir_data.py:172]: I think I don't hate this anymore. This object is not long-lived but that's fine.

File: conda/core/subdir_data.py [hunk 8]
```diff
@@ -395,30 +408,38 @@ def _read_pickled(self, etag, mod_stamp):
...
             return None
 
+        def _pickle_valid_checks():
+            yield "_url", _pickled_state.get("_url"), self.url_w_credentials
+            yield "_schannel", _pickled_state.get("_schannel"), self.channel.canonical_name
+            yield "_add_pip", _pickled_state.get("_add_pip"), context.add_pip_as_python_dependency
+            yield "_mod", _pickled_state.get("_mod"), state.get("_mod")
+            yield "_etag", _pickled_state.get("_etag"), state.get("_etag")
+            yield "_pickle_version", _pickled_state.get("_pickle_version"), REPODATA_PICKLE_VERSION
+            yield "fn", _pickled_state.get("fn"), self.repodata_fn
```
Comment (dholth) [inline conda/core/subdir_data.py:418]: Helpful for debugging; you can ask for all checks even if first one fails

File: conda/core/subdir_data.py [hunk 9]
```diff
@@ -395,30 +408,38 @@ def _read_pickled(self, etag, mod_stamp):
...
+    def _process_raw_repodata_str(self, raw_repodata_str, state={}):
+        """
+        state contains information that was previously in-band in raw_repodata_str.
+        """
         json_obj = json.loads(raw_repodata_str or "{}")
-        return self._process_raw_repodata(json_obj)
+        return self._process_raw_repodata(json_obj, state=state)
 
-    def _process_raw_repodata(self, repodata):
+    def _process_raw_repodata(self, repodata, state={}):
```
Comment (dholth) [inline conda/core/subdir_data.py:442]: mutable dict should not be default

File: conda/core/subdir_data.py [hunk 10]
```diff
@@ -808,10 +603,24 @@ def cache_fn_url(url, repodata_fn=REPODATA_FN):
...
 
-def add_http_value_to_dict(resp, http_key, d, dict_key):
-    value = resp.headers.get(http_key)
-    if value:
-        d[dict_key] = value
+def fetch_repodata_remote_request(url, etag, mod_stamp, repodata_fn=REPODATA_FN):
+    # this function should no longer be used by conda but is kept for API stability
+    warnings.warn(
+        "fetch_repodata_remote_request",
+        PendingDeprecationWarning,
```
Comment (dholth) [inline conda/core/subdir_data.py:610]: todo write the automatic deprecation until version x.y function

File: conda/gateways/repodata/__init__.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+
+class CondaRepoInterface(RepoInterface):
+    def __init__(self, url: str, repodata_fn: str | None, **kwargs) -> None:
+        self._url = url
+        self._repodata_fn = repodata_fn
+
+        self._log = logging.getLogger(__name__)
+        self._stderrlog = logging.getLogger("conda.stderrlog")
+
+    def repodata(self, state: dict) -> str | None:
```
Comment (dholth) [inline conda/gateways/repodata/__init__.py:67]: this is copied from the old fetch_repodata_remote_request as much as possible

File: conda/gateways/repodata/__init__.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+        last_modified = state.get("_mod")
+        if etag:
+            headers["If-None-Match"] = etag
+        if last_modified:
+            headers["If-Modified-Since"] = last_modified
+        filename = self._repodata_fn
+
+        url = join_url(self._url, filename)
+
+        with conda_http_errors(self._url, filename):
```
Comment (dholth) [inline conda/gateways/repodata/__init__.py:84]: this with statement factors out the non-reusable baroque error handling from the old function

File: conda/gateways/repodata/__init__.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+            if self._log.isEnabledFor(logging.DEBUG):
+                self._log.debug(stringify(resp, content_max_len=256))
+            resp.raise_for_status()
+
+        if resp.status_code == 304:
+            raise Response304ContentUnchanged()
+
+        def maybe_decompress(filename, resp_content):
+            return ensure_text_type(
+                bz2.decompress(resp_content) if filename.endswith(".bz2") else resp_content
```
Comment (dholth) [inline conda/gateways/repodata/__init__.py:96]: todo zstd :)

File: conda/gateways/repodata/__init__.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+def _add_http_value_to_dict(resp, http_key, d, dict_key):
+    value = resp.headers.get(http_key)
+    if value:
+        d[dict_key] = value
+
+
+@contextmanager
+def conda_http_errors(url, repodata_fn):
+    """
+    Use in a with: statement to translate requests exceptions to conda ones.
```
[the remark this answers is not in the document]
Comment (dholth) [reply conda/gateways/repodata/__init__.py:121]: also nearly identical to the old fetch_repodata_remote_request

File: conda/gateways/repodata/__init__.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+    except (ConnectionError, HTTPError) as e:
+        status_code = getattr(e.response, "status_code", None)
+        if status_code in (403, 404):
+            if not url.endswith("/noarch"):
+                log.info(
+                    "Unable to retrieve repodata (response: %d) for %s",
+                    status_code,
+                    url + "/" + repodata_fn,
+                )
+                raise RepodataIsEmpty(
```
Comment (dholth) [inline conda/gateways/repodata/__init__.py:171]: the old one did "return None" here, which we can't do in the with handler.

Comment (dholth) [reply conda/gateways/repodata/__init__.py:171]: hidden in here is a simple "retry if current_repodata.json" is not available behavior.

**Review by travishathaway — APPROVED** [review]

**Review by travishathaway — CHANGES_REQUESTED** [review]
I forgot to remind you to add a news snippet. Thanks.

**Review by travishathaway — DISMISSED** [review]

File: tests/requirements.txt (hunk 4 above)
Comment (jezdez) [reply tests/requirements.txt:18]: Have we in between gotten these versions on defaults?

**Review by jezdez — CHANGES_REQUESTED** [review]
A few nits to fix, but this looks almost ready to merge.

File: tests/requirements.txt (hunk 4 above)
Comment (dholth) [reply tests/requirements.txt:18]: No. Recent jinja2 and werkzeug are available though. https://anaconda.org/main/flask

**Comment by dholth** [comment]
This should be tested against "file:/// repositories without current_repodata.json". Either this or the repodata/jlap branch retries too many times in that case.

File: tests/requirements.txt (hunk 4 above)
Comment (dholth) [reply tests/requirements.txt:18]: I may be able to defer this until the next branch.

**Comment by dholth** [comment]
> A few nits to fix, but this looks almost ready to merge.

@jezdez adjusted.

**Comment by dholth** [comment]
These tests have developed a frustrating fault related to dll paths? If you have any insight please send help, don't see how anything here relates to shell activation.

File: conda/core/subdir_data.py [hunk 16]
```diff
@@ -235,6 +240,27 @@ def iter_records(self):
...
+            # json.loads(state_path.read_text()) and better that open("rb")
+            with state_path.open("r") as s:
+                state = json.load(s)
+            return state
+        except (json.JSONDecodeError, OSError):
+            log.debug("Could not load state", exc_info=True)
+            return {}
+
+    def _save_state(self, state: dict):
+        return pathlib.Path(self.cache_path_state).write_text(json.dumps(state, indent=True))
```
Comment (jezdez) [inline conda/core/subdir_data.py:262]: Do we need to do any exception handling for this, so the API works the same as with `_load_state`?

Comment (jezdez) [reply conda/core/subdir_data.py:262]: Alternatively, you could move the error handling out of `_load_state` into where it's called.

Comment (dholth) [reply conda/core/subdir_data.py:262]: If you can't save this, it's as bad as not being able to save repodata.json

Comment (jezdez) [reply conda/core/subdir_data.py:262]: Okay.

**Review by jezdez — APPROVED** [review]
Let's get this merged!

File: conda/core/subdir_data.py (hunk 10 above)
Comment (jezdez) [reply conda/core/subdir_data.py:629]: Oh wait, this warning needs to be updated to be end-user compatible :D

**Review by jezdez — CHANGES_REQUESTED** [review]

**Comment by jezdez** [comment]
> What is an end-user-compatible message? Will PendingDeprecationWarning messages be shown without a special developer flag?
> 
> > Both of DeprecationWarning and PendingDeprecationWarning are suppressed by default.
> > When using -Wd or -Xdev option, both are enabled.
> 
> If a third party (not conda which no longer uses it) calls this function, and enables deprecation warnings, then they will see the message.

I shouldn't have used “end-user” that was misleading, but “developer-friendly” instead. In my experience, the deprecation warnings are very useful for developers of 3rd party tools who depend on conda, to show them during development or testing. IIRC, pytest shows them by default? It's a way to literally print on their terminal that they should stop using that particular code, since we can't assume they read every commit of conda.

**Review by jezdez — APPROVED** [review]
🚀

File: conda/core/subdir_data.py [hunk 17]
```diff
@@ -205,6 +196,13 @@ def url_w_repodata_fn(self):
     def cache_path_json(self):
         return self.cache_path_base + ("1" if context.use_only_tar_bz2 else "") + ".json"
 
+    @property
+    def cache_path_state(self):
+        """
+        Out-of-band etag and other state needed by the RepoInterface.
+        """
+        return self.cache_path_base + ".state.json"
```
[the remark this answers is not in the document]
Comment (dholth) [reply conda/core/subdir_data.py:208]: wolfv and I have a thread on conda slack

[the remark this answers is not in the document]
Comment (jezdez) [reply conda/core/subdir_data.py:208]: Yeah, I understand that, an alternative would be to implement a highly performant conda cache library that can be used from both?

[the remark this answers is not in the document]
Comment (dholth) [reply conda/core/subdir_data.py:208]: Also Navigator, which we broke not with the cache but with our download messages.

**Pull request merged**